File: opendaw/packages/lib/dsp/src/waveshaper.ts
import {int, Unhandled} from "@opendaw/lib-std"

// https://www.desmos.com/calculator/04tpdtpkfy
export namespace Waveshaper {
    export type Equation = "hardclip" | "cubicSoft" | "tanh" | "sigmoid" | "arctan" | "asymmetric"

    export const Equations: ReadonlyArray<Equation> = ["hardclip", "cubicSoft", "tanh", "sigmoid", "arctan", "asymmetric"]

    export const apply = (x: number, equation: Equation): number => {
        switch (equation) {
            case "hardclip":
                return x < -1.0 ? -1.0 : x > 1.0 ? 1.0 : x
            case "cubicSoft": {
                const cx = x < -1.0 ? -1.0 : x > 1.0 ? 1.0 : x
                return (3.0 * cx - cx * cx * cx) * 0.5
            }
            case "tanh":
                return Math.tanh(x)
            case "sigmoid":
                return Math.sign(x) * (1.0 - Math.exp(-Math.abs(x)))
            case "arctan":
                return (2.0 / Math.PI) * Math.atan(x)
            case "asymmetric":
                if (x >= 0.0) return x / (1.0 + x)
                if (x >= -2.0 / 3.0) return x
                if (x >= -1.0) {
                    const t = 3.0 * (x + 1.0)
                    return t * t * (2.0 - t) / 3.0 - 1.0
                }
                return -1.0
            default:
                return Unhandled(equation)
        }
    }

    export const process = (audio: [Float32Array, Float32Array], equation: Equation, fromIndex: int, toIndex: int): void => {
        const [l, r] = audio
        const {sign, exp, abs, atan, tanh, PI} = Math
        const TWO_OVER_PI = 2.0 / PI
        switch (equation) {
            case "hardclip":
                for (let i = fromIndex; i < toIndex; i++) {
                    l[i] = l[i] < -1.0 ? -1.0 : l[i] > 1.0 ? 1.0 : l[i]
                    r[i] = r[i] < -1.0 ? -1.0 : r[i] > 1.0 ? 1.0 : r[i]
                }
                return
            case "cubicSoft":
                for (let i = fromIndex; i < toIndex; i++) {
                    const lx = l[i] < -1.0 ? -1.0 : l[i] > 1.0 ? 1.0 : l[i]
                    const rx = r[i] < -1.0 ? -1.0 : r[i] > 1.0 ? 1.0 : r[i]
                    l[i] = (3.0 * lx - lx * lx * lx) * 0.5
                    r[i] = (3.0 * rx - rx * rx * rx) * 0.5
                }
                return
            case "tanh":
                for (let i = fromIndex; i < toIndex; i++) {
                    l[i] = tanh(l[i])
                    r[i] = tanh(r[i])
                }
                return
            case "sigmoid":
                for (let i = fromIndex; i < toIndex; i++) {
                    l[i] = sign(l[i]) * (1.0 - exp(-abs(l[i])))
                    r[i] = sign(r[i]) * (1.0 - exp(-abs(r[i])))
                }
                return
            case "arctan":
                for (let i = fromIndex; i < toIndex; i++) {
                    l[i] = TWO_OVER_PI * atan(l[i])
                    r[i] = TWO_OVER_PI * atan(r[i])
                }
                return
            case "asymmetric":
                for (let i = fromIndex; i < toIndex; i++) {
                    if (l[i] >= 0.0) {
                        l[i] = l[i] / (1.0 + l[i])
                    } else if (l[i] < -1.0) {
                        l[i] = -1.0
                    } else if (l[i] < -2.0 / 3.0) {
                        const tl = 3.0 * (l[i] + 1.0)
                        l[i] = tl * tl * (2.0 - tl) / 3.0 - 1.0
                    }
                    if (r[i] >= 0.0) {
                        r[i] = r[i] / (1.0 + r[i])
                    } else if (r[i] < -1.0) {
                        r[i] = -1.0
                    } else if (r[i] < -2.0 / 3.0) {
                        const tr = 3.0 * (r[i] + 1.0)
                        r[i] = tr * tr * (2.0 - tr) / 3.0 - 1.0
                    }
                }
                return
            default:
                return Unhandled(equation)
        }
    }
}

File: opendaw/crates/dsp/src/waveshaper.rs
//! Waveshaping transfer functions, a port of lib-dsp `Waveshaper`. Each equation maps one sample through a
//! non-linearity; `process` shapes a stereo block in place. `f32` (the audio path), mirroring the TS `number`
//! math. https://www.desmos.com/calculator/04tpdtpkfy

use core::f32::consts::PI;

/// The transfer functions, in the TS declaration order (`Waveshaper.Equations`) — the `equation` string field
/// resolves to one of these by name.
#[derive(Clone, Copy, PartialEq, Eq, Default, Debug)]
pub enum Equation {
    #[default]
    HardClip,
    CubicSoft,
    Tanh,
    Sigmoid,
    Arctan,
    Asymmetric
}

impl Equation {
    /// Resolve the box's `equation` string field. An empty / unknown string falls back to `Tanh`, mirroring the
    /// TS processor (`value === "" ? "tanh" : value`).
    pub fn from_name(name: &str) -> Self {
        match name {
            "hardclip" => Equation::HardClip,
            "cubicSoft" => Equation::CubicSoft,
            "tanh" => Equation::Tanh,
            "sigmoid" => Equation::Sigmoid,
            "arctan" => Equation::Arctan,
            "asymmetric" => Equation::Asymmetric,
            _ => Equation::Tanh
        }
    }

    /// Map one sample through the transfer function. Mirrors `Waveshaper.apply`.
    #[inline]
    pub fn apply(self, x: f32) -> f32 {
        match self {
            Equation::HardClip => x.clamp(-1.0, 1.0),
            Equation::CubicSoft => {
                let cx = x.clamp(-1.0, 1.0);
                (3.0 * cx - cx * cx * cx) * 0.5
            }
            Equation::Tanh => libm::tanhf(x),
            Equation::Sigmoid => sign(x) * (1.0 - libm::expf(-libm::fabsf(x))),
            Equation::Arctan => (2.0 / PI) * libm::atanf(x),
            Equation::Asymmetric => {
                if x >= 0.0 {
                    x / (1.0 + x)
                } else if x < -1.0 {
                    -1.0
                } else if x < -2.0 / 3.0 {
                    let t = 3.0 * (x + 1.0);
                    t * t * (2.0 - t) / 3.0 - 1.0
                } else {
                    x
                }
            }
        }
    }
}

/// Shape a stereo block in place over `[from, to)`. Mirrors `Waveshaper.process`.
pub fn process(left: &mut [f32], right: &mut [f32], equation: Equation, from: usize, to: usize) {
    for sample in from..to {
        left[sample] = equation.apply(left[sample]);
        right[sample] = equation.apply(right[sample]);
    }
}

/// `Math.sign`: -1 / 0 / +1. (`f32::signum` returns ±1 for zero, which would differ from the TS `sign`.)
#[inline]
fn sign(x: f32) -> f32 {
    if x > 0.0 {
        1.0
    } else if x < 0.0 {
        -1.0
    } else {
        0.0
    }
}

#[cfg(test)]
mod tests {
    use super::{process, Equation};

    #[test]
    fn names_resolve_and_unknown_falls_back_to_tanh() {
        assert_eq!(Equation::from_name("hardclip"), Equation::HardClip);
        assert_eq!(Equation::from_name("asymmetric"), Equation::Asymmetric);
        assert_eq!(Equation::from_name(""), Equation::Tanh);
        assert_eq!(Equation::from_name("nonsense"), Equation::Tanh);
    }

    #[test]
    fn hardclip_clamps_to_unit_range() {
        assert_eq!(Equation::HardClip.apply(2.0), 1.0);
        assert_eq!(Equation::HardClip.apply(-2.0), -1.0);
        assert_eq!(Equation::HardClip.apply(0.5), 0.5);
    }

    #[test]
    fn tanh_is_odd_and_saturates() {
        let big = Equation::Tanh.apply(10.0);
        assert!(big > 0.999, "saturates toward +1");
        assert!((Equation::Tanh.apply(-0.3) + Equation::Tanh.apply(0.3)).abs() < 1e-6, "odd symmetry");
    }

    #[test]
    fn arctan_is_bounded_by_one() {
        assert!(Equation::Arctan.apply(1e6) < 1.0 && Equation::Arctan.apply(1e6) > 0.999);
    }

    #[test]
    fn asymmetric_treats_positive_and_negative_differently() {
        let pos = Equation::Asymmetric.apply(1.0); // 1/(1+1) = 0.5
        assert!((pos - 0.5).abs() < 1e-6);
        assert_eq!(Equation::Asymmetric.apply(-2.0), -1.0); // clamps the far-negative tail
        assert!((Equation::Asymmetric.apply(-0.5) - (-0.5)).abs() < 1e-6); // linear in the near-negative region
    }

    #[test]
    fn process_shapes_a_stereo_block_in_place() {
        let mut left = [2.0f32, -2.0, 0.25];
        let mut right = [-3.0f32, 0.5, 4.0];
        process(&mut left, &mut right, Equation::HardClip, 0, 3);
        assert_eq!(left, [1.0, -1.0, 0.25]);
        assert_eq!(right, [-1.0, 0.5, 1.0]);
    }

    #[test]
    fn process_respects_the_range_bounds() {
        let mut left = [2.0f32, 2.0, 2.0];
        let mut right = [2.0f32, 2.0, 2.0];
        process(&mut left, &mut right, Equation::HardClip, 1, 2); // only index 1
        assert_eq!(left, [2.0, 1.0, 2.0]);
        assert_eq!(right, [2.0, 1.0, 2.0]);
    }
}
